src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -25,252 +26,1010 @@
 #include "gc/shenandoah/shenandoahHeapRegionSet.hpp"
 #include "gc/shenandoah/shenandoahMarkingContext.inline.hpp"
 #include "logging/logStream.hpp"
 #include "memory/resourceArea.hpp"
 #include "runtime/orderAccess.hpp"
 
-ShenandoahFreeSet::ShenandoahFreeSet(ShenandoahHeap* heap, size_t max_regions) :
-  _heap(heap),
-  _mutator_free_bitmap(max_regions, mtGC),
-  _collector_free_bitmap(max_regions, mtGC),
-  _max(max_regions)
+static const char* partition_name(ShenandoahFreeSetPartitionId t) {
+  switch (t) {
+    case NotFree: return "NotFree";
+    case Mutator: return "Mutator";
+    case Collector: return "Collector";
+    default: return "Unrecognized";
+  }
+}
+
+size_t ShenandoahSimpleBitMap::count_leading_ones(ssize_t start_idx) const {
+  assert((start_idx >= 0) && (start_idx < _num_bits), "precondition");
+  size_t array_idx = start_idx >> LogBitsPerWord;
+  uintx element_bits = _bitmap[array_idx];
+  uintx bit_number = start_idx & right_n_bits(LogBitsPerWord);
+  uintx omit_mask = right_n_bits(bit_number);
+  uintx mask = ((uintx) 0 - 1) & ~omit_mask;
+  if ((element_bits & mask) == mask) {
+    size_t counted_ones = BitsPerWord - bit_number;
+    return counted_ones + count_leading_ones(start_idx - counted_ones);
+  } else {
+    // Return number of consecutive ones starting with the_bit and including more significant bits.
+    uintx aligned = element_bits >> bit_number;
+    uintx complement = ~aligned;;
+    return count_trailing_zeros<uintx>(complement);
+  }
+}
+
+size_t ShenandoahSimpleBitMap::count_trailing_ones(ssize_t last_idx) const {
+  assert((last_idx >= 0) && (last_idx < _num_bits), "precondition");
+  size_t array_idx = last_idx >> LogBitsPerWord;
+  uintx element_bits = _bitmap[array_idx];
+  uintx bit_number = last_idx & right_n_bits(LogBitsPerWord);
+  // All ones from bit 0 to the_bit
+  uintx mask = right_n_bits(bit_number + 1);
+  if ((element_bits & mask) == mask) {
+    size_t counted_ones = bit_number + 1;
+    return counted_ones + count_trailing_ones(last_idx - counted_ones);
+  } else {
+    // Return number of consecutive ones starting with the_bit and including less significant bits
+    uintx aligned = element_bits << (BitsPerWord - (bit_number + 1));
+    uintx complement = ~aligned;
+    return count_leading_zeros<uintx>(complement);
+  }
+}
+
+bool ShenandoahSimpleBitMap::is_forward_consecutive_ones(ssize_t start_idx, ssize_t count) const {
+  assert((start_idx >= 0) && (start_idx < _num_bits), "precondition: start_idx: " SSIZE_FORMAT ", count: " SSIZE_FORMAT,
+         start_idx, count);
+  assert(start_idx + count <= (ssize_t) _num_bits, "precondition");
+  size_t array_idx = start_idx >> LogBitsPerWord;
+  uintx bit_number = start_idx & right_n_bits(LogBitsPerWord);
+  uintx element_bits = _bitmap[array_idx];
+  uintx bits_to_examine  = BitsPerWord - bit_number;
+  element_bits >>= bit_number;
+  uintx complement = ~element_bits;
+  uintx trailing_ones;
+  if (complement) {
```

Comment:
complement is not a bool type. If you meant to say 'if (complement != 0)' then say so, to make it easier for the reader. Same applies for a few other places.

Reply:
Thanks.  Fixed.